qpext/cmd/qpext/main.go

```diff
@@ -308,39 +311,51 @@ func (sc *ScrapeConfigurations) handleStats(w http.ResponseWriter, r *http.Reque
 func main() {
 	zapLogger := logger.InitializeLogger()
 	mux := http.NewServeMux()
-	ctx := context.Background()
+	ctx, cancel := context.WithCancel(context.Background())
+	aggregateMetricsPort := os.Getenv(QueueProxyAggregatePrometheusMetricsPortEnvVarKey)
 	sc := NewScrapeConfigs(
 		zapLogger,
 		QueueProxyMetricsPort,
 		os.Getenv(KServeContainerPrometheusMetricsPortEnvVarKey),
 		os.Getenv(KServeContainerPrometheusMetricsPathEnvVarKey),
 	)
 	mux.HandleFunc(`/metrics`, sc.handleStats)
-	l, err := net.Listen("tcp", fmt.Sprintf(":%v", os.Getenv(QueueProxyAggregatePrometheusMetricsPortEnvVarKey)))
+	l, err := net.Listen("tcp", fmt.Sprintf(":%v", aggregateMetricsPort))
 	if err != nil {
 		zapLogger.Error("error listening on status port", zap.Error(err))
 		return
 	}
 
-	defer l.Close()
-
+	errCh := make(chan error)
 	go func() {
+		zapLogger.Info(fmt.Sprintf("Starting stats server on port %v", aggregateMetricsPort))
 		if err = http.Serve(l, mux); err != nil {
-			zapLogger.Error("error serving aggregate metrics", zap.Error(err))
-			fmt.Println(err)
-			select {
-			case <-ctx.Done():
-				// We are shutting down already, don't trigger SIGTERM
-				return
-			default:
-			}
+			errCh <- fmt.Errorf("stats server failed to serve: %w", err)
+		}
+	}()
+
+	go func() {
+		if err := sharedmain.Main(); err != nil {
+			errCh <- err
 		}
+		// sharedMain exited without error which means graceful shutdown due to SIGTERM / SIGINT signal
+		// Attempt a graceful shutdown of stats server
+		cancel()
 	}()
```

Comment:
I am not sure if this is needed as `sharedmain.Main()` spawns the go channel under the hood
https://github.com/knative/serving/blob/main/pkg/queue/sharedmain/main.go. Were you able to replicate the issue?

Reply:
Shared.Main [will block the main thread](https://github.com/knative/serving/blob/60ee8de2aaf125e5cf0008e225f7f3983d7af2a3/pkg/queue/sharedmain/main.go#L281C1-L306C3), so we have to spawn a new go routine.